### DISCUSSION TITLE
"Known compatibility issue"

### DISCUSSION THREAD

**Discussion by FIFCC** [revyos/revyos#83] [Announcements]
Document the currently known issues regarding compatibility.

**Comment by FIFCC** [comment]
Milk-V Meles cannot display or change resolution on certain 4K monitors.

Reference [issue81](https://github.com/revyos/revyos/issues/81)

**Comment by FIFCC** [comment]
LicheePi4A image flashing problem

In the following environment, flashing the image will encounter the problem of not being able to recognize the board

Hardware: cpu: 9950x motherboard: b650e
System: fedora41, kernel version 6.12.4

In the same system environment, using cpu: 12400 and motherboard: b660 can perform normal flashing operations.
![image](https://github.com/user-attachments/assets/61b0ba06-b6b2-4a84-b5c6-ca0b1a8f4a00)

**Reply by FIFCC** [reply]
Please see this issue for details[issue87](https://github.com/revyos/revyos/issues/87)

**Comment by FIFCC** [comment]
The video encoding and decoding not working in the new LicheePi 4A image with the 6.6 kernel is a known issue.

**Reply by FIFCC** [reply]
Device: LicheePi 4A
Image Version:[20241229](https://mirror.iscas.ac.cn/revyos/extra/images/lpi4a/20241229/)

**Comment by FIFCC** [comment]
When the wireless module is unavailable, check whether the onboard wireless card is AIC8800 (11ax, 2.4GHz) or RTL8723DS (11n, 2.4GHz). The AIC8800 is more likely to fail to function.

**Reply by kagura114** [reply]
The AIC8800 WiFi module is available starting from RevyOS 20250123. Upgrading to the latest kernel via GitHub Actions or updating the system will enable WiFi functionality. However, issues with MMC2 (see #101 ) persist, causing WiFi to become unavailable after reboot. To resolve this, restart the board by unplugging the power instead of performing a direct reboot.

**Comment by FIFCC** [comment]
After booting and running Meles from a TF card for a period of time (around one week), the file system may become inaccessible. Restarting the system can resolve the issue.

**Comment by FIFCC** [comment]
The current 1229 version of the LPI4A image cannot boot on the LPi4A with 8GB RAM + 32GB eMMC. The 16GB RAM + 128GB eMMC version is not affected by this issue for now. For more details, please refer to this [issue](https://github.com/revyos/revyos/issues/91).

**Comment by FIFCC** [comment]
To launch the Chromium browser via the command line, please use the following command:  
`chromium -no-sandbox -use-gl=egl`

**Comment by KevinMX** [comment]
Some USB (UVC) cameras won't work on TH1520 based boards, e.g. Milk-V Meles.

See this issue for some known not working models: https://github.com/revyos/revyos/issues/78

**Comment by RP38** [comment]
Hi, I am using 5.10 revyos for beaglev ahead. 

I tried to use the NPU (ax3386) with the AI tutorials available on [sipeed](https://wiki.sipeed.com/hardware/en/lichee/th1520/lpi4a/8_application.html#MobilenetV2), but I notice a very high latency when using the NPU. it takes 1 minutes to load the model (graph_create), but once the model is loaded, It seems to work as expected (much higher fps than c910). Did anyone encounter this behavior ? When I run a model on the CPU, I do not have the problem so I guess it must be a problem with the NPU driver or SHL library.

With gdb, I managed to identify that the problem is likely due to shl_pnna.h library which we don't have the sources [(th1520 sources aren't available)](https://github.com/XUANTIE-RV/csi-nn2/issues/17).

**Reply by KamijoToma** [reply]
Hi! That might because the [demo program](https://github.com/zhangwm-pt/lpi4a-example/blob/main/classification/mobilenetv2/main.cpp#L156) always try to load the original model `hhb_out/hhb.bm`, however a jit version is available at `hhb_out/shl.hhb.bm`. It can be generated by running `hhb_out/hhb_jit hhb_out/hhb.bm` or just running the demo program. Once it is generated, you can replace the `hhb.bm` argument in your `system()` call with `shl.hhb.bm`.

**Reply by KamijoToma** [reply]
You can see the difference of time consumption between using `hhb.bm` and `shl.hhb.bm`:
```
debian@revyos-lpi4a:~/resnet/hhb_out$ sudo time ./hhb_runtime hhb.bm data.0.bin
INFO: NNA clock:425412 [kHz]
INFO: Heap :anonymous (0x2)
INFO: Heap :dmabuf (0x2)
INFO: Heap :unified (0x5)
WARNING: Mapping to the on chip ram failed (128 > 0), continuing...
FATAL: Importing 150528 bytes of CPU memory has failed (wrong memory alignment)
Run graph execution time: 18.04168ms, FPS=55.42722

=== tensor info ===
shape: 1 3 224 224
data pointer: 0x2d16c100

=== tensor info ===
shape: 1 1000
data pointer: 0x3f80c87000
The max_value of output: 16.002394
The min_value of output: -4.726096
The mean_value of output: -0.000497
The std_value of output: 4.274896
 ============ top5: ===========
283: 16.002394
281: 9.783847
285: 7.793912
287: 7.462256
282: 6.550203
93.30user 0.69system 1:34.10elapsed 99%CPU (0avgtext+0avgdata 342812maxresident)k
0inputs+35616outputs (0major+85088minor)pagefaults 0swaps
debian@revyos-lpi4a:~/resnet/hhb_out$ sudo time ./hhb_runtime shl.hhb.bm data.0.bin
INFO: NNA clock:425412 [kHz]
INFO: Heap :anonymous (0x2)
INFO: Heap :dmabuf (0x2)
INFO: Heap :unified (0x5)
WARNING: Mapping to the on chip ram failed (128 > 0), continuing...
FATAL: Importing 150528 bytes of CPU memory has failed (wrong memory alignment)
Run graph execution time: 17.93862ms, FPS=55.74565

=== tensor info ===
shape: 1 3 224 224
data pointer: 0x3f8192f010

=== tensor info ===
shape: 1 1000
data pointer: 0x3f8506b000
The max_value of output: 16.002394
The min_value of output: -4.726096
The mean_value of output: -0.000497
The std_value of output: 4.274896
 ============ top5: ===========
283: 16.002394
281: 9.783847
285: 7.793912
287: 7.462256
282: 6.550203
0.03user 0.11system 0:00.15elapsed 94%CPU (0avgtext+0avgdata 50392maxresident)k
0inputs+24outputs (0major+5975minor)pagefaults 0swaps
debian@revyos-lpi4a:~/resnet/hhb_out$
```

**Reply by RP38** [reply]
Thank you, it is indeed much faster. Any info on what jit does ?

**Reply by KamijoToma** [reply]
according this function in `hhb_out/main.c`:

```
void *create_graph(char *params_path) {
    int binary_size;
    char *params = get_binary_from_file(params_path, &binary_size);
    if (params == NULL) {
        return NULL;
    }

    char *suffix = params_path + (strlen(params_path) - 7);
    if (strcmp(suffix, ".params") == 0) {
        // create general graph
        return csinn_(params);
    }

    suffix = params_path + (strlen(params_path) - 3);
    if (strcmp(suffix, ".bm") == 0) {
        struct shl_bm_sections *section = (struct shl_bm_sections *)(params + 4128);
        if (section->graph_offset) {
            return csinn_import_binary_model(params);
        } else {
            return csinn_(params + section->params_offset * 4096);
        }
    } else {
        return NULL;
    }
}
```

If we pass `model.params`, it will call `csinn_()` in `model.c`, which is a very large function(3000+ on resnet) that manually creates tensors and links them. however `shl.hhb.bm` causes it to call `csinn_import_binary_model()`, which maybe the accelerated jit version to load model params into NPU.
